### PR TITLE
Document GameOptions struct + management code

### PR DESCRIPTION
**Pull request by lhearachel**
## Renamed Files

* `include/struct_defs/struct_020279FC.h` -> `include/struct_defs/game_options.h` (do we want to maintain the `struct_` prefix here?)
* `include/unk_020279FC.h` -> `include/game_options.h`
* `src/unk_020279FC.c` -> `src/game_options.c`

## Renamed Structs and Fields

* `UnkStruct_020279FC` -> `GameOptions`
* `UnkStruct_020279FC::unk_00_0` -> `GameOptions::textSpeed`
* `UnkStruct_020279FC::unk_00_4` -> `GameOptions::audioMode`
* `UnkStruct_020279FC::unk_00_6` -> `GameOptions::battleStyle`
* `UnkStruct_020279FC::unk_00_7` -> `GameOptions::battleScene`
* `UnkStruct_020279FC::unk_02_0` -> `GameOptions::buttonMode`
* `UnkStruct_020279FC::unk_02_2` -> `GameOptions::frameStyle`
* `UnkStruct_020279FC::` -> `GameOptions::_padding`

## Renamed Functions

* `sub_020279FC` -> `GameOptions_Init`
* `sub_02027A10` -> `GameOptions_Copy`
* `sub_02027A1C` -> `GameOptions_SetDefaultValues`
* `sub_02027A68` -> `GameOptions_UpdateGlobalButtonMode`
* `sub_02027AA4` -> `GameOptions_GetTextSpeed`
* `sub_02027AAC` -> `GameOptions_SetTextSpeed`
* `sub_02027AC0` -> `GameOptions_GetTextSpeedReductionFactor`
* `sub_02027ADC` -> `GameOptions_GetAudioMode`
* `sub_02027AE4` -> `GameOptions_SetAudioMode`
* `sub_02027AF8` -> `GameOptions_GetBattleScene`
* `sub_02027B00` -> `GameOptions_SetBattleScene`
* `sub_02027B14` -> `GameOptions_GetBattleStyle`
* `sub_02027B1C` -> `GameOptions_SetBattleStyle`
* `sub_02027B30` -> `GameOptions_GetButtonMode`
* `sub_02027B38` -> `GameOptions_SetButtonMode`
* `sub_02027B50` -> `GameOptions_GetFrameStyle`
* `sub_02027B58` -> `GameOptions_SetFrameStyle`

## New Enums and Typedefs

* `TextSpeed`: defines permitted values for `GameOptions::textSpeed`
* `AudioMode`: defines permitted values for `GameOptions::audioMode`
* `BattleScene`: defines permitted values for `GameOptions::battleScene`
* `BattleStyle`: defines permitted values for `GameOptions::battleStyle`
* `ButtonMode`: defines permitted values for `GameOptions::buttonMode`
* `FrameStyle`: defines permitted values for `GameOptions::frameStyle`

### CODE REVIEW

File: src/game_options.c [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    options->textSpeed = newVal;
+}
+
+u8 GameOptions_GetTextSpeedReductionFactor (const GameOptions * options)
+{
+    TextSpeed rawTextSpeedValue = GameOptions_GetTextSpeed(options);
+    switch (rawTextSpeedValue) {
+        case TEXTSPEED_SLOW:        return 8;
+        case TEXTSPEED_MID:         return 4;
+        default:                    return 1;
```
Comment (luckytyphlosion) [inline src/game_options.c:74]: return statements should be on new line indented

File: src/game_options.c [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+void GameOptions_SetDefaultValues (GameOptions * options)
+{
+    MI_CpuFill8(options, 0, sizeof(GameOptions));
+
+    options->textSpeed      = TEXTSPEED_MID;
+    options->audioMode      = AUDIOMODE_STEREO;
+    options->battleScene    = BATTLESCENE_ON;
+    options->battleStyle    = BATTLESTYLE_SHIFT;
+    options->buttonMode     = BUTTONMODE_NORMAL;
+    options->frameStyle     = FRAMESTYLE_01;
```
Comment (luckytyphlosion) [inline src/game_options.c:36]: do not align

File: src/game_options.c [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            break;
+    }
+}
+
+TextSpeed GameOptions_GetTextSpeed (const GameOptions * options)
+{
+    return options->textSpeed;
+}
+
+void GameOptions_SetTextSpeed (GameOptions * options, TextSpeed newVal)
```
[the remark this answers is not in the document]
Comment (luckytyphlosion) [reply src/game_options.c:63]: actually didn't realize these were enums, they should say e.g. `enum TextSpeed` (do this for all)

File: include/game_options.h [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+#include "struct_defs/game_options.h"
+#include "struct_decls/struct_021C0794_decl.h"
+
+typedef enum {
+    TEXTSPEED_SLOW = 0,
+    TEXTSPEED_MID,
+    TEXTSPEED_FAST,
+    TEXTSPEED_NUM,
+} TextSpeed;
```
Comment (luckytyphlosion) [inline include/game_options.h:12]: convert to non-typedef (do for all enums used)

File: include/struct_defs/game_options.h [hunk 5]
```diff
@@ -0,0 +1,52 @@
+#ifndef POKEPLATINUM_STRUCT_GAME_OPTIONS_H
+#define POKEPLATINUM_STRUCT_GAME_OPTIONS_H
+
+/**
+ * Managed struct of configurable options for a game which are customizable
+ * by the player.
+ */
+typedef struct {
+    u16 textSpeed      : 4;     ///< Controls the speed at which text renders
```
Comment (luckytyphlosion) [inline include/struct_defs/game_options.h:9]: Do not align between textSpeed and colon (do for all instances)

File: include/struct_defs/game_options.h [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+                                ///< If `buttonMode == 1`, then `Start` will
+                                ///< be treated as `X`.
+                                ///< If `buttonMode == 2`, then `L` will be
+                                ///< treated as `A`.
+
+    u16 frameStyle     : 5;     ///< Designates which frame style to load and
+                                ///< render when displaying a scrolling message
+                                ///< box. Platinum comes with 20 of these.
+
+    u16 _padding       : 1;     ///< Padding alignment to 16 bits.
```
Comment (luckytyphlosion) [inline include/struct_defs/game_options.h:49]: Do not prefix with `_` (names prefixed with `_` are reserved under C standard)

File: include/struct_defs/game_options.h [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+ * by the player.
+ */
+typedef struct {
+    u16 textSpeed      : 4;     ///< Controls the speed at which text renders
+                                ///< in scrolling message frames. The raw value
+                                ///< must be translated into a reduction factor.
+                                ///< Supports values up to 15, but all values
+                                ///< greater than 1 are treated as equivalents.
+
+    u16 audioMode      : 2;     ///< Controls the method by which audio is
```
Comment (luckytyphlosion) [inline include/struct_defs/game_options.h:15]: Is it possible to use the enums here? Don't remember if possible to set enum type, but don't bother if not matching

File: include/struct_defs/game_options.h [hunk 8]
```diff
@@ -0,0 +1,52 @@
+#ifndef POKEPLATINUM_STRUCT_GAME_OPTIONS_H
```
Comment (luckytyphlosion) [inline include/struct_defs/game_options.h:1]: Move this stuff in the same file as game_options.h (do not use struct_defs, that's a temp folder that originated from the automated struct splitting)

File: src/game_options.c [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+{
+    options->textSpeed = newVal;
+}
+
+u8 GameOptions_GetTextSpeedReductionFactor (const GameOptions * options)
+{
+    TextSpeed rawTextSpeedValue = GameOptions_GetTextSpeed(options);
```
Comment (luckytyphlosion) [inline src/game_options.c:70]: Maybe think of better terminology (need to differentiate between textSpeed enum value (FAST/MID/SLOW) and text speed delay value)

Comment (luckytyphlosion) [reply src/game_options.c:70]: LetterDelay is probably a better term here, or even just GetTextSpeedDelay (so rename `rawTextSpeedValue ` to `textSpeed` and function name to either `GameOptions_GetTextSpeedLetterDelay` or `GameOptions_GetTextSpeedDelay`)

File: src/game_options.c [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+    options->textSpeed      = TEXTSPEED_MID;
+    options->audioMode      = AUDIOMODE_STEREO;
+    options->battleScene    = BATTLESCENE_ON;
+    options->battleStyle    = BATTLESTYLE_SHIFT;
+    options->buttonMode     = BUTTONMODE_NORMAL;
+    options->frameStyle     = FRAMESTYLE_01;
+}
+
+void GameOptions_UpdateGlobalButtonMode (UnkStruct_021C0794 * saveData, ButtonMode newButtonMode)
```
Comment (luckytyphlosion) [inline src/game_options.c:39]: Do we know if this is `newButtonMode` for a fact, even though `options->buttonMode` isn't set?

**Review by luckytyphlosion — CHANGES_REQUESTED** [review]

**Comment by PikalaxALT** [comment]
Is this PR abandoned?

**Pull request closed without merging**